### PR TITLE
Composite and binary gutenberg dependency

### PR DESCRIPTION
**Pull request by oguzkocer**
This PR removes `libs/gutenberg-mobile` submodule by changing `react-native-bridge` to be a binary dependency. It also allows using a local path for `gutenberg-mobile` repo to be used as a composite build. These workflows are made available by the changes in this Gutenberg PR: https://github.com/WordPress/gutenberg/pull/29488

_**Note that this PR can not be merged until the linked Gutenberg PR and a currently WIP gutenberg-mobile PR (which will add auto-publishing of `react-native-bridge` and `react-native-aztec` libraries) are merged.**_

#### Binary dependency

* This is now the default way `react-native-bridge` is added as a dependency.
* `ext.gutenbergMobileVersion` property is used to download `react-native-bridge` from our own Bintray.
* We need to add the `hermes` aar for binary builds as it's not available through any remote repositories. This might be something we can improve in the future, but for the moment, it's a necessary step.

#### Composite build

* Similar to other composite builds, we only need to add the path for `localGutenbergMobilePath` to the `local-builds.gradle` file.
* Composite builds always work as if the previously used `BUILD_GUTENBERG_FROM_SOURCE` is set to `true`. This not only simplifies the build files, but also makes the workflows more straightforward.
* We no longer need to add the extra `react-native-xx` projects as this will now be done through `react-native-bridge`.
* Hermes is available through the `react-native-bridge` project.

---

#### Binary dependency

* We simply need to build & run the project and test the Gutenberg blocks, ideally all of them.

#### Composite build

* Checkout [mokagio/publish-to-bintray-with-script](https://github.com/wordpress-mobile/gutenberg-mobile/tree/mokagio/publish-to-bintray-with-script) branch from your local `gutenberg-mobile` project
* Add your local path for `gutenberg-mobile` project as `localGutenbergMobilePath` to `local-builds.gradle` file (create the file by copying `local-builds.gradle-example` file if you don't have one)
* Follow the instructions from [gutenberg-mobile README](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/README.md) to get your metro server running. (Main steps being, installing latest npm, running `npm install` and `npm run start:reset` commands from your `gutenberg-mobile` project)
* Run the app and test the Gutenberg blocks, ideally all of them.
* Make changes to JS files from your `gutenberg-mobile` project and verify that it's reflected in the app. (This is very similar to how you'd do it in current `develop`, only difference being, you should now change your local `gutenberg-mobile` repo instead of the `libs/gutenberg-mobile` folder which is no longer available)

---

## _Extras_

* It'd be great to test this PR in Windows. The project hasn't been working in Windows since we switched away from Jitpack for `react-native-bridge`, but with this change, the binary build should just work. I'll try and verify this when I can.
* I'd love to do some benchmarks comparing the current `develop` and this branch for different type of builds to see how much of a speed difference we'll gain with the binary dependency and whether there is any significant downside to the composite build for the `BUILD_GUTENBERG_FROM_SOURCE` case.

## Remaining tasks

- [x] Remove CircleCI tasks related to gutenberg-mobile submodule
- [ ] Figure out how string translations will be handled cc @loremattei 
- [ ] Auto publishing `react-native-bridge` from `gutenberg-mobile` to Bintray

### CODE REVIEW

File: libs/editor/WordPressEditor/build.gradle
Comment (mokagio) [inline libs/editor/WordPressEditor/build.gradle]:
```suggestion
    // This dependency will be substituted if the `local-builds.gradle` file contains
```

File: libs/editor/WordPressEditor/build.gradle
Comment (mokagio) [inline libs/editor/WordPressEditor/build.gradle]: If possible, it would be good to have a link to a page with Google's rationale.

**Review by mokagio — COMMENTED** [review]
image

🎉 🎉 🎉 

---

`assembleVanillaDebug` succeeded in both configurations. I run the app using the composite build, I always forget to call `npm run start:reset`, but once I got that going, it worked great. I posted [this](https://amokafullofstuff.wordpress.com/2021/03/03/test-post-from-android-composite-build/) from the Emulator running with it. 

Leaving this as a "Comment" and not "Approve" to have one last run once the Gutenberg PR is merged in.

File: libs/editor/WordPressEditor/build.gradle
Comment (mchowning) [inline libs/editor/WordPressEditor/build.gradle]: I feel like we may have already discussed this and I'm just forgetting, but does the `BUILD_GUTENBERG_FROM_SOURCE` variable still have a purpose? I would really like to remove it if possible.

Comment (oguzkocer) [reply libs/editor/WordPressEditor/build.gradle]: [This](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java#L535-L537) is why we still need to pass this. If we can rework that somehow, we can remove this config field.

Comment (mchowning) [reply libs/editor/WordPressEditor/build.gradle]: Could we tie this to the same logic that determines if we're doing a local build? In other words, this would be true if and only if we were locally building Gutenberg?

Comment (oguzkocer) [reply libs/editor/WordPressEditor/build.gradle]: @mchowning Do you mean do this from the Gutenberg repo? Otherwise, that's exactly how it currently works. The `buildGutenbergFromSource` is not something developers should modify, it is set depending on whether they have the `localGutenbergMobilePath` property in their `local-builds.gradle` file or not.

File: local-builds.gradle-example
Comment (mchowning) [inline local-builds.gradle-example]: Would it make more sense to comment all these out by default? The odds that someone would have all three of these setup at the proper path and actually want to build all three of these from source seems to be virtually 0, so it seems like a non-ideal starting point. It also virtually guarantees that the first build after copying this file will fail. The failure is relatively clear, but it does present a stumbling block, particularly if someone copies the file over not fully understanding what it is doing. Just a thought.

File: libs/editor/WordPressEditor/build.gradle
Comment (mchowning) [reply libs/editor/WordPressEditor/build.gradle]: Ah I missed that 🤦 :

```
def buildGutenbergFromSource = gradle.ext.includedBuildGutenbergMobilePath != null
```

Thanks for explaining. 🙇

Comment (mchowning) [reply libs/editor/WordPressEditor/build.gradle]: I don't feel strongly, but do you think it would be worth renaming this variable and config field to something like `gutenbergIncludedBuild` and `GUTENBERG_INCLUDED_BUILD` just to make that even more obvious?

File: libs/editor/WordPressEditor/build.gradle
Comment (oguzkocer) [reply libs/editor/WordPressEditor/build.gradle]: Here is a Stackoverflow answer by "Xavier Ducrohet" who is the "Android SDK Tech Lead - Google Inc." according to their profile: https://stackoverflow.com/a/20715155

@mokagio @mchowning I don't typically like to refer to stackoverflow from code comments, but I guess this one is OK since it's from an official source. What do you think?

I am also open to the idea of rewording this comment if you can think of a better way to explain this.

File: libs/editor/WordPressEditor/build.gradle
Comment (oguzkocer) [reply libs/editor/WordPressEditor/build.gradle]: I am not particularly happy with this flag either, let me see if I can read this property from `react-native-bridge` instead.

File: libs/editor/WordPressEditor/build.gradle
Comment (oguzkocer) [reply libs/editor/WordPressEditor/build.gradle]: I am not sure if I agree, but added in cc4f290b63a378f6773db74370ec03796426a594.

File: local-builds.gradle-example
Comment (oguzkocer) [reply local-builds.gradle-example]: Whenever this file is copied, I expect at least one of the libraries to be used as a composite build, so I thought the developer would decide which ones to comment out and which ones to keep. Having said that, commenting out them sounds like a reasonable default. Updated in cc4f290b63a378f6773db74370ec03796426a594.

File: libs/editor/WordPressEditor/build.gradle
Comment (mokagio) [reply libs/editor/WordPressEditor/build.gradle]: >  I don't typically like to refer to stackoverflow from code comments, but I guess this one is OK since it's from an official source. What do you think?

I think that's okay in this instance. Even if the comment was reworded, I'd still look at it and ask "what's the source?". I would have been nice if Google had an official page about this but 🤷‍♂️ 


P.S. _That comment is from 2013_ 😱 They really must not care about this kind of problem / don't see it as the right problem to tackle.

**Comment by mokagio** [comment]
@oguzkocer this branch looks like the final version of the first iteration of our work for the Gutenberg binary + composite build, right?

If that's the case, should we remove the original WIP branch, [`gutenberg-composite-build`](https://github.com/wordpress-mobile/WordPress-Android/tree/gutenberg-composite-build)?

If the answer is yes, feel free to remove it. I closed all my old PRs depending on it.

File: libs/editor/WordPressEditor/build.gradle
Comment (oguzkocer) [reply libs/editor/WordPressEditor/build.gradle]: I've changed the wording a bit and added the link in 178ac0b5fc746df907aa2b00860dbd037b6de46c.

> P.S. That comment is from 2013 😱 They really must not care about this kind of problem / don't see it as the right problem to tackle.

I think it might just be due to its complication. They might believe that it's best to publish each component separately. I haven't given it enough thought, but I think there is merit to it. I'll see if we can upload this binary somewhere and read it from there, but I don't want it blocking the merge.

File: libs/editor/WordPressEditor/build.gradle
Comment (oguzkocer) [reply libs/editor/WordPressEditor/build.gradle]: I've reworked this in https://github.com/oguzkocer/gutenberg/commit/d9a4a70ed6deaa29f0a35c55cc6f565259e1c689 and b45beeb29110a9e5dc36aa1d9df8b5bd49a3c029. We no longer need to pass any arguments to `react-native-bridge` as it'll decide whether to include the bundle or not.

**Review by mchowning — APPROVED** [review]
Tested and looks good! 👍

**Pull request merged**
